### PR TITLE
Add check in readme for invalid plugin name header

### LINKED ISSUE #397: Missing invalid plugin name header check in Readme Check
This is the check implemented in wordpress.org readme validator. It is missing in our Readme Check.

In Readme check, following check is 
```php
// Fatal errors.
if ( empty( $readme->name ) || isset( $readme->warnings['invalid_plugin_name_header'] ) ) {
	$errors[] = sprintf(
		/* translators: 1: 'Plugin Name' section title, 2: 'Plugin Name' */
		__( 'We cannot find a plugin name in your readme. Plugin names look like: %1$s. Please change %2$s to reflect the actual name of your plugin.', 'wporg-plugins' ),
		'<code>=== Plugin Name ===</code>',
		'<code>Plugin Name</code>'
	);
}
```

Ref: https://github.com/WordPress/wordpress.org/blob/trunk/wordpress.org/public_html/wp-content/plugins/plugin-directory/readme/class-validator.php#L73

### PR DESCRIPTION
**Pull request by ernilambar**
Fixes https://github.com/WordPress/plugin-check/issues/397

* Add check for invalid plugin name header in readme.
* ERROR is raised if plugin name in the readme is invalid.

### CODE REVIEW

**Comment by ernilambar** [comment]
cc: @swissspidy @felixarntz @mukeshpanchal27

File: tests/phpunit/testdata/plugins/test-plugin-plugin-readme-errors-invalid-name/readme.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+=== Plugin Name ===
```
Comment (swissspidy) [inline tests/phpunit/testdata/plugins/test-plugin-plugin-readme-errors-invalid-name/readme.txt:1]: And why is this invalid? 🤔

Comment (ernilambar) [reply tests/phpunit/testdata/plugins/test-plugin-plugin-readme-errors-invalid-name/readme.txt:1]: https://github.com/WordPress/wordpress.org/blob/trunk/wordpress.org/public_html/wp-content/plugins/plugin-directory/readme/class-parser.php#L227 `if ( 'plugin name' == strtolower( $this->name ) ) {
			$this->warnings['invalid_plugin_name_header'] = true; ...`

Comment (swissspidy) [reply tests/phpunit/testdata/plugins/test-plugin-plugin-readme-errors-invalid-name/readme.txt:1]: Ah 💡

File: includes/Checker/Checks/Plugin_Readme_Check.php [hunk 2]
```diff
@@ -87,6 +90,28 @@ protected function check_files( Check_Result $result, array $files ) {
...
+	 * @param Parser       $parser      The Parser object.
+	 */
+	private function check_name( Check_Result $result, string $readme_file, Parser $parser ) {
+		if ( empty( $parser->name ) || isset( $parser->warnings['invalid_plugin_name_header'] ) ) {
+			$message = sprintf(
+				/* translators: 1: 'Plugin Name' section title, 2: 'Plugin Name' */
+				__( 'We cannot find a plugin name in your readme. Plugin names look like: %1$s. Please change %2$s to reflect the actual name of your plugin.', 'plugin-check' ),
+				"'=== Plugin Name ==='",
+				"'Plugin Name'"
+			);
```
Comment (swissspidy) [inline includes/Checker/Checks/Plugin_Readme_Check.php]: Maybe let's add two separate error messages, one for the `empty()` case and one for when the plugin name is "Plugin Name".

**Comment by ernilambar** [comment]
Unit test is failing because Parser package is not updated in this PR. This PR should be merged after this https://github.com/WordPress/plugin-check/pull/395

**Comment by swissspidy** [comment]
> Unit test is failing because Parser package is not updated in this PR. This PR should be merged after this #395

Can you do the same update in this PR?

File: tests/phpunit/tests/Checker/Checks/Plugin_Readme_Check_Tests.php [hunk 3]
```diff
@@ -31,6 +31,30 @@ public function test_run_with_errors_no_readme() {
 		$this->assertEquals( 'no_plugin_readme', $warnings['readme.txt'][0][0][0]['code'] );
 	}
 
+	public function test_run_with_errors_invalid_name() {
```
Comment (swissspidy) [inline tests/phpunit/tests/Checker/Checks/Plugin_Readme_Check_Tests.php:34]: We need a test case for `empty_plugin_name` too

**Comment by ernilambar** [comment]
* Test case added for empty plugin name.
* Now Testing is failed because of `unset( $warnings['invalid_plugin_name_header'] );`. We have done this in another PR. Because we have not removed this item from warnings list, test is failing.

**Comment by swissspidy** [comment]
Can we make the same change here so that this doesn't depend on another PR?

**Review by swissspidy — APPROVED** [review]

File: tests/phpunit/tests/Checker/Checks/Plugin_Readme_Check_Tests.php [hunk 4]
```diff
@@ -31,6 +31,54 @@ public function test_run_with_errors_no_readme() {
...
+		// Check for invalid name error.
+		$this->assertArrayHasKey( 0, $errors['readme.txt'] );
+		$this->assertArrayHasKey( 0, $errors['readme.txt'][0] );
+		$this->assertArrayHasKey( 'code', $errors['readme.txt'][0][0][0] );
+		$this->assertEquals( 'invalid_plugin_name', $errors['readme.txt'][0][0][0]['code'] );
+	}
+
+	public function test_run_with_errors_empty_name() {
+		$readme_check  = new Plugin_Readme_Check();
+		$check_context = new Check_Context( UNIT_TESTS_PLUGIN_DIR . 'test-plugin-plugin-readme-errors-empty-name/load.php' );
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/Checker/Checks/Plugin_Readme_Check_Tests.php:60]: If we only remove the plugin name from the readme file, the parser does not return an error. 🤔 Maybe it's a bug in the parser itself? I tried this to remove the `no_license` error from these tests.

File: tests/phpunit/testdata/plugins/test-plugin-plugin-readme-errors-empty-name/readme.txt
```
Contributors:      plugin-check
Requires at least: 6.0
Tested up to:      6.1
Requires PHP:      5.6
Stable tag:        n.e.x.t
License:           GPLv2 or later
License URI:       https://www.gnu.org/licenses/gpl-2.0.html
Tags:              testing, security

A simple readme file for testing.
```

File: includes/Checker/Checks/Plugin_Readme_Check.php [hunk 5]
```diff
@@ -87,6 +90,36 @@ protected function check_files( Check_Result $result, array $files ) {
...
+	 *
+	 * @param Check_Result $result      The Check Result to amend.
+	 * @param string       $readme_file Readme file.
+	 * @param Parser       $parser      The Parser object.
+	 */
+	private function check_name( Check_Result $result, string $readme_file, Parser $parser ) {
+		if ( isset( $parser->warnings['invalid_plugin_name_header'] ) ) {
+			$message = sprintf(
+				/* translators: 1: 'Plugin Name' section title, 2: 'Plugin Name' */
+				__( 'Plugin names look like: "%1$s". Please change "%2$s" to reflect the actual name of your plugin.', 'plugin-check' ),
```
Comment (mukeshpanchal27) [inline includes/Checker/Checks/Plugin_Readme_Check.php]:
```suggestion
				__( 'Plugin name look like: "%1$s". Please change "%2$s" to reflect the actual name of your plugin.', 'plugin-check' ),
```

File: tests/phpunit/tests/Checker/Checks/Plugin_Readme_Check_Tests.php (hunk 4 above)
Comment (ernilambar) [reply tests/phpunit/tests/Checker/Checks/Plugin_Readme_Check_Tests.php:60]: Yes. That's the bug in the Parser. In your above example Parser takes `Contributors:      plugin-check` as `name` and continues to parse other as headers.

Comment (mukeshpanchal27) [reply tests/phpunit/tests/Checker/Checks/Plugin_Readme_Check_Tests.php:60]: In that case could you please open ticket on meta for tracking?

Comment (ernilambar) [reply tests/phpunit/tests/Checker/Checks/Plugin_Readme_Check_Tests.php:60]: Meta ticket for improving parsing of plugin name: https://meta.trac.wordpress.org/ticket/7428

**Review by mukeshpanchal27 — APPROVED** [review]

**Pull request merged**